File: api/services/storage_service.py
import json
import uuid
from pathlib import Path
from datetime import datetime


DATA_DIR = Path(__file__).resolve().parent.parent / "data"


class StorageService:
    def __init__(self):
        DATA_DIR.mkdir(exist_ok=True)
        self._ensure_files()

    def _ensure_files(self):
        for name in ["areas.json", "alerts.json", "history.json"]:
            path = DATA_DIR / name
            if not path.exists():
                path.write_text("[]")

    def _read(self, name: str) -> list:
        path = DATA_DIR / name
        if path.exists():
            return json.loads(path.read_text())
        return []

    def _write(self, name: str, data):
        (DATA_DIR / name).write_text(json.dumps(data, indent=2, ensure_ascii=False))

    # ── AREAS ──

    def get_areas(self) -> list:
        return self._read("areas.json")

    def get_area(self, area_id: str) -> dict | None:
        for a in self.get_areas():
            if a.get("id") == area_id:
                return a
        return None

    def save_area(self, area: dict) -> dict:
        areas = self.get_areas()
        if not area.get("id"):
            area["id"] = f"AREA_{uuid.uuid4().hex[:8].upper()}"
            area["createdAt"] = datetime.now().isoformat()
            areas.append(area)
        else:
            for i, a in enumerate(areas):
                if a.get("id") == area["id"]:
                    area["updatedAt"] = datetime.now().isoformat()
                    areas[i] = {**a, **area}
                    break
            else:
                areas.append(area)
        self._write("areas.json", areas)
        return area

    def update_area(self, area_id: str, updates: dict) -> dict | None:
        areas = self.get_areas()
        for i, a in enumerate(areas):
            if a.get("id") == area_id:
                areas[i] = {**a, **updates, "updatedAt": datetime.now().isoformat()}
                self._write("areas.json", areas)
                return areas[i]
        return None

    def delete_area(self, area_id: str) -> bool:
        areas = self.get_areas()
        new_areas = [a for a in areas if a.get("id") != area_id]
        if len(new_areas) == len(areas):
            return False
        self._write("areas.json", new_areas)
        return True

    # ── ALERTS ──

    def get_alerts(self) -> list:
        return self._read("alerts.json")

    def save_alert(self, alert: dict) -> dict:
        alerts = self.get_alerts()
        if not alert.get("id"):
            alert["id"] = f"ALT_{uuid.uuid4().hex[:8].upper()}"
            alert["createdAt"] = datetime.now().isoformat()
            alerts.insert(0, alert)
        else:
            for i, a in enumerate(alerts):
                if a.get("id") == alert["id"]:
                    alerts[i] = {**a, **alert}
                    break
        self._write("alerts.json", alerts)
        return alert

    def update_alert(self, alert_id: str, updates: dict) -> dict | None:
        alerts = self.get_alerts()
        for i, a in enumerate(alerts):
            if a.get("id") == alert_id:
                alerts[i] = {**a, **updates, "updatedAt": datetime.now().isoformat()}
                self._write("alerts.json", alerts)
                return alerts[i]
        return None

    # ── HISTORY ──

    def get_history(self, area_id: str = None) -> list:
        history = self._read("history.json")
        if area_id:
            history = [h for h in history if h.get("areaId") == area_id]
        return history

    def add_history(self, entry: dict) -> dict:
        history = self._read("history.json")
        entry["id"] = f"HIST_{uuid.uuid4().hex[:8].upper()}"
        entry["timestamp"] = datetime.now().isoformat()
        history.insert(0, entry)
        self._write("history.json", history)
        return entry

File: site/geo/js/ui/services/StorageService.js
/**
 * STORAGE SERVICE
 * Gerencia persistencia local (LocalStorage)
 */

class StorageService {
    constructor() {
        this.prefix = APP_CONFIG.STORAGE.PREFIX;
        this.areasKey = this.prefix + APP_CONFIG.STORAGE.AREAS_KEY;
        this.alertsKey = this.prefix + APP_CONFIG.STORAGE.ALERTS_KEY;
        this.historyKey = this.prefix + APP_CONFIG.STORAGE.HISTORY_KEY;
        this.settingsKey = this.prefix + APP_CONFIG.STORAGE.SETTINGS_KEY;
        this._initialized = false;
    }

    init() {
        if (this._initialized) return;
        const areas = this.getAreas();
        if (areas.length === 0) {
            this._loadDemoData();
        }
        this._initialized = true;
        console.log('[StorageService] Inicializado com sucesso');
    }

    _loadDemoData() {
        if (typeof DEMO_DATA !== 'undefined') {
            this.saveAreas(DEMO_DATA.areas || []);
            this.saveAlerts(DEMO_DATA.alerts || []);
            this.saveHistory(DEMO_DATA.history || []);
            console.log('[StorageService] Dados de demonstracao carregados');
        }
    }

    // ==================== AREAS ====================

    getAreas() {
        try {
            const data = localStorage.getItem(this.areasKey);
            return data ? JSON.parse(data) : [];
        } catch (e) {
            console.error('[StorageService] Erro ao ler areas:', e);
            return [];
        }
    }

    saveAreas(areas) {
        try {
            localStorage.setItem(this.areasKey, JSON.stringify(areas));
        } catch (e) {
            console.error('[StorageService] Erro ao salvar areas:', e);
        }
    }

    getArea(id) {
        const areas = this.getAreas();
        return areas.find(a => a.id === id) || null;
    }

    saveArea(area) {
        const areas = this.getAreas();
        if (!area.id) {
            area.id = this._generateId();
            area.createdAt = new Date().toISOString();
            areas.push(area);
        } else {
            const index = areas.findIndex(a => a.id === area.id);
            if (index !== -1) {
                area.updatedAt = new Date().toISOString();
                areas[index] = { ...areas[index], ...area };
            } else {
                areas.push(area);
            }
        }
        this.saveAreas(areas);
        return area;
    }

    updateArea(id, updates) {
        const areas = this.getAreas();
        const index = areas.findIndex(a => a.id === id);
        if (index === -1) return null;
        areas[index] = { ...areas[index], ...updates, updatedAt: new Date().toISOString() };
        this.saveAreas(areas);
        return areas[index];
    }

    deleteArea(id) {
        const areas = this.getAreas();
        const filtered = areas.filter(a => a.id !== id);
        if (filtered.length === areas.length) return false;
        this.saveAreas(filtered);
        return true;
    }

    // ==================== ALERTS ====================

    getAlerts() {
        try {
            const data = localStorage.getItem(this.alertsKey);
            return data ? JSON.parse(data) : [];
        } catch (e) {
            console.error('[StorageService] Erro ao ler alertas:', e);
            return [];
        }
    }

    saveAlerts(alerts) {
        try {
            localStorage.setItem(this.alertsKey, JSON.stringify(alerts));
        } catch (e) {
            console.error('[StorageService] Erro ao salvar alertas:', e);
        }
    }

    saveAlert(alert) {
        const alerts = this.getAlerts();
        if (!alert.id) {
            alert.id = this._generateId('ALT');
            alert.createdAt = new Date().toISOString();
            alerts.unshift(alert);
        } else {
            const index = alerts.findIndex(a => a.id === alert.id);
            if (index !== -1) {
                alert.updatedAt = new Date().toISOString();
                alerts[index] = { ...alerts[index], ...alert };
            }
        }
        this.saveAlerts(alerts);
        return alert;
    }

    updateAlert(id, updates) {
        const alerts = this.getAlerts();
        const index = alerts.findIndex(a => a.id === id);
        if (index === -1) return null;
        alerts[index] = { ...alerts[index], ...updates, updatedAt: new Date().toISOString() };
        this.saveAlerts(alerts);
        return alerts[index];
    }

    // ==================== HISTORY ====================

    getHistory() {
        try {
            const data = localStorage.getItem(this.historyKey);
            return data ? JSON.parse(data) : [];
        } catch (e) {
            console.error('[StorageService] Erro ao ler historico:', e);
            return [];
        }
    }

    saveHistory(history) {
        try {
            localStorage.setItem(this.historyKey, JSON.stringify(history));
        } catch (e) {
            console.error('[StorageService] Erro ao salvar historico:', e);
        }
    }

    addHistory(entry) {
        const history = this.getHistory();
        entry.id = this._generateId('HIST');
        entry.timestamp = entry.timestamp || new Date().toISOString();
        history.unshift(entry);
        this.saveHistory(history);
        return entry;
    }

    getHistoryByArea(areaId) {
        return this.getHistory().filter(h => h.areaId === areaId);
    }

    // ==================== HELPERS ====================

    _generateId(prefix = '') {
        const timestamp = Date.now().toString(36).toUpperCase();
        const random = Math.random().toString(36).substring(2, 6).toUpperCase();
        return prefix ? `${prefix}_${timestamp}_${random}` : `${timestamp}_${random}`;
    }

    clearAll() {
        localStorage.removeItem(this.areasKey);
        localStorage.removeItem(this.alertsKey);
        localStorage.removeItem(this.historyKey);
        this._initialized = false;
    }

    exportGeoJSON() {
        const areas = this.getAreas();
        return {
            type: 'FeatureCollection',
            features: areas.map(area => ({
                type: 'Feature',
                properties: {
                    id: area.id, nome: area.nome, categoria: area.categoria,
                    prioridade: area.prioridade, status: area.status,
                    areaHa: area.areaHa || 0, createdAt: area.createdAt
                },
                geometry: area.geojson ? area.geojson.geometry : null
            })).filter(f => f.geometry)
        };
    }

    importGeoJSON(geojson) {
        if (!geojson || geojson.type !== 'FeatureCollection') throw new Error('GeoJSON invalido');
        const areas = geojson.features.map(feature => ({
            id: this._generateId(), nome: feature.properties.nome || 'Area Importada',
            categoria: feature.properties.categoria || 'preservacao',
            prioridade: feature.properties.prioridade || 'media',
            status: feature.properties.status || 'ativo', frequency: 'manual',
            geojson: feature, areaHa: feature.properties.areaHa || 0,
            createdAt: new Date().toISOString()
        }));
        const existing = this.getAreas();
        this.saveAreas([...existing, ...areas]);
        return areas;
    }
}

const storageService = new StorageService();
window.storageService = storageService;
